src/eins-hwinfo.c

```diff
@@ -0,0 +1,417 @@
+  guint64 used = 0;
+  guint64 available = 0;
+
+  info = g_file_query_filesystem_info (file,
+                                       G_FILE_ATTRIBUTE_FILESYSTEM_SIZE ","
+                                       G_FILE_ATTRIBUTE_FILESYSTEM_USED ","
+                                       G_FILE_ATTRIBUTE_FILESYSTEM_FREE,
+                                       NULL /* cancellable */,
+                                       error);
+  if (info != NULL)
+    {
+      total = g_file_info_get_attribute_uint64 (info,
+                                                G_FILE_ATTRIBUTE_FILESYSTEM_SIZE);
+      used = g_file_info_get_attribute_uint64 (info,
+                                               G_FILE_ATTRIBUTE_FILESYSTEM_USED);
+      available = g_file_info_get_attribute_uint64 (info,
+                                                    G_FILE_ATTRIBUTE_FILESYSTEM_FREE);
+    }
+
+  return g_variant_new ("(uuu)",
+                        round_to_nearest_gigabyte (total),
+                        round_to_nearest_gigabyte (used),
+                        round_to_nearest_gigabyte (available));
+}
+
+static gboolean
+is_mounted (GFile *file)
+{
+  g_autoptr(GFileInfo) info = g_file_query_info (file,
+                                                 G_FILE_ATTRIBUTE_UNIX_IS_MOUNTPOINT,
+                                                 G_FILE_QUERY_INFO_NOFOLLOW_SYMLINKS,
+                                                 NULL, NULL);
+
+  return info != NULL &&
+    g_file_info_get_attribute_boolean (info, G_FILE_ATTRIBUTE_UNIX_IS_MOUNTPOINT);
+}
+
+/**
+ * @data: G_SOURCE_REMOVE or G_SOURCE_CONTINUE
+ */
+static gboolean
+record_disk_space (gpointer data)
+{
+  g_autoptr(GFile) root = g_file_new_for_path ("/");
+  g_autoptr(GFile) extra = g_file_new_for_path ("/var/endless-extra");
+  g_autoptr(GError) error = NULL;
+  GVariant *root_payload =
+    eins_hwinfo_get_disk_space_for_partition (root, &error);
+
+  if (error != NULL)
+    {
+      g_warning ("Couldn't get disk space for /: %s", error->message);
+      g_clear_error (&error);
+    }
+
+  emtr_event_recorder_record_event (emtr_event_recorder_get_default (),
+                                    SYSROOT_DISK_SPACE_EVENT,
+                                    g_steal_pointer (&root_payload));
+
+  if (is_mounted (extra))
+    {
+      GVariant *extra_payload =
+        eins_hwinfo_get_disk_space_for_partition (extra, &error);
+
+      if (error != NULL)
+        g_warning ("Couldn't get disk space for /var/endless-extra: %s", error->message);
+
+      emtr_event_recorder_record_event (emtr_event_recorder_get_default (),
+                                        EXTRA_DISK_SPACE_EVENT,
+                                        g_steal_pointer (&extra_payload));
+    }
+
+  return GPOINTER_TO_INT (data);
+}
+
+GVariant *
+eins_hwinfo_get_ram_size (void)
+{
+  glibtop_mem mem;
+
+  glibtop_get_mem (&mem);
+  return g_variant_new_uint32 (round_to_nearest_gigabyte (mem.total));
```

Comment:
I... er... prefer RAM in megabytes? Maybe even mibibytes. Things like UMA apertures and other BIOS options can steal non-integer number of gigabytes from the system.

Reply:
good point!

Reply:
Changed both to power-of-2; changed this to mebibytes. There's an argument that we should always round towards 0. I implemented round-to-nearest because that's what I implemented in the winstaller, but can't remember why that's what I implemented there.